File: chrome-plugin/src/language/validator-messages.test.ts
import { readFileSync } from "node:fs";
import { describe, expect, it } from "vitest";
import type { ValidationError } from "./analysis-validator";
import { validateCoreBatch } from "./analysis-validator";
import { tokenize } from "./segmenter";

interface ValidatorMessageFixture {
  schemaVersion: number;
  coveredMessageSubstrings: string[];
  cases: ValidatorMessageCase[];
}

interface ValidatorMessageCase {
  id: string;
  sentence: {
    id: string;
    text: string;
  };
  raw: unknown;
  expected: ValidationError[];
}

const fixture = JSON.parse(
  readFileSync(
    new URL("../../../shared-fixtures/validator-messages.json", import.meta.url),
    "utf8",
  ),
) as ValidatorMessageFixture;

function actualErrors(testCase: ValidatorMessageCase): ValidationError[] {
  const sentence = {
    sentenceId: testCase.sentence.id,
    text: testCase.sentence.text,
    tokens: tokenize(testCase.sentence.text),
  };
  const result = validateCoreBatch(testCase.raw, [sentence], "validator-messages-fixture");
  expect(result.ok, `fixture case ${testCase.id} must be invalid`).toBe(false);
  return result.ok ? [] : result.errors;
}

describe("shared validator message fixture", () => {
  it("has an unambiguous versioned schema", () => {
    expect(fixture.schemaVersion).toBe(1);
    expect(fixture.cases.length).toBeGreaterThan(0);
    expect(new Set(fixture.cases.map(({ id }) => id)).size).toBe(fixture.cases.length);
    expect(fixture.cases.every(({ expected }) => expected.length > 0)).toBe(true);
  });

  it.each(fixture.cases)("matches the complete ordered errors for $id", (testCase) => {
    expect(actualErrors(testCase)).toEqual(testCase.expected);
  });

  it("keeps covered message substrings closed over all fixture errors", () => {
    const messages = fixture.cases.flatMap(actualErrors).map(({ message }) => message);

    expect(
      fixture.coveredMessageSubstrings.filter(
        (substring) => !messages.some((message) => message.includes(substring)),
      ),
      "every declared substring must be represented by an actual fixture error",
    ).toEqual([]);
    expect(
      messages.filter(
        (message) =>
          !fixture.coveredMessageSubstrings.some((substring) => message.includes(substring)),
      ),
      "every actual fixture error must belong to a declared covered message family",
    ).toEqual([]);
  });
});

File: intellij-plugin/src/test/kotlin/dev/codetui/englishsyntax/language/ValidatorMessagesTest.kt
package dev.codetui.englishsyntax.language

import dev.codetui.englishsyntax.contract.FixtureLoader
import dev.codetui.englishsyntax.domain.SentenceInput
import kotlinx.serialization.json.Json
import kotlinx.serialization.json.JsonArray
import kotlinx.serialization.json.JsonObject
import kotlinx.serialization.json.jsonArray
import kotlinx.serialization.json.jsonObject
import kotlinx.serialization.json.jsonPrimitive
import kotlin.test.Test
import kotlin.test.assertEquals
import kotlin.test.assertTrue

class ValidatorMessagesTest {
  private val fixture = Json.parseToJsonElement(FixtureLoader.text("validator-messages.json")).jsonObject

  @Test
  fun `fixture schema is versioned and unambiguous`() {
    val cases = fixture.getValue("cases").jsonArray
    val ids = cases.map { it.jsonObject.getValue("id").jsonPrimitive.content }

    assertEquals(1, fixture.getValue("schemaVersion").jsonPrimitive.content.toInt())
    assertTrue(cases.isNotEmpty())
    assertEquals(ids.size, ids.toSet().size)
    assertTrue(cases.all {
      val testCase = it.jsonObject
      val accepted = testCase.getValue("accepted").jsonPrimitive.content.toBooleanStrict()
      val expected = testCase.getValue("expected").jsonArray
      if (accepted) expected.isEmpty() else expected.isNotEmpty()
    })
  }

  @Test
  fun `Kotlin validator matches every complete ordered TypeScript error list`() {
    fixture.getValue("cases").jsonArray.forEach { element ->
      val testCase = element.jsonObject
      assertEquals(expectedErrors(testCase), actualErrors(testCase), "fixture case ${testCase.id()}")
    }
  }

  @Test
  fun `covered message substrings are closed over fixture errors`() {
    val substrings = fixture.getValue("coveredMessageSubstrings").jsonArray.map { it.jsonPrimitive.content }
    val messages = fixture.getValue("cases").jsonArray
      .flatMap { actualErrors(it.jsonObject) }
      .map { it.message }

    assertEquals(
      emptyList(),
      substrings.filter { substring -> messages.none { it.contains(substring) } },
      "every declared substring must be represented by an actual fixture error",
    )
    assertEquals(
      emptyList(),
      messages.filter { message -> substrings.none { message.contains(it) } },
      "every actual fixture error must belong to a declared covered message family",
    )
  }

  private fun actualErrors(testCase: JsonObject) = run {
    val sentenceJson = testCase.getValue("sentence").jsonObject
    val text = sentenceJson.getValue("text").jsonPrimitive.content
    val sentence = SentenceInput(
      sentenceId = sentenceJson.getValue("id").jsonPrimitive.content,
      text = text,
      tokens = tokenize(text),
    )
    val result = validateCoreBatch(testCase.getValue("raw"), listOf(sentence), "validator-messages-fixture")
    val accepted = testCase.getValue("accepted").jsonPrimitive.content.toBooleanStrict()

    assertEquals(accepted, result.ok, "fixture case ${testCase.id()} acceptance")
    result.errors
  }

  private fun expectedErrors(testCase: JsonObject) = testCase.getValue("expected").jsonArray.map { element ->
    val expected = element.jsonObject
    expected.getValue("path").jsonPrimitive.content to expected.getValue("message").jsonPrimitive.content
  }

  private fun JsonObject.id() = getValue("id").jsonPrimitive.content

  private fun List<dev.codetui.englishsyntax.domain.ValidationError>.asPairs() = map { it.path to it.message }

  private fun assertEquals(
    expected: List<Pair<String, String>>,
    actual: List<dev.codetui.englishsyntax.domain.ValidationError>,
    message: String,
  ) = assertEquals(expected, actual.asPairs(), message)
}
